### PR TITLE
Flutter iOS Interactive Keyboard: Update Scroll View Widget

### PR DESCRIPTION
**Pull request by Matt2D**
This PR address the movement aspect of the flutter interactive keyboard. Updates the scroll view widget, adding the interactive behavior enum case for keyboardDismissBehavior. This is a desired behavior of the keyboard that has not yet been implemented.

Design Document:
https://docs.google.com/document/d/1-T7_0mSkXzPaWxveeypIzzzAdyo-EEuP5V84161foL4/edit?pli=1

Issues Address:
https://github.com/flutter/flutter/issues/57609

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 1]
```diff
@@ -38,6 +39,9 @@ enum ScrollViewKeyboardDismissBehavior {
   /// `onDrag` means that the [ScrollView] will dismiss an on-screen keyboard
   /// when a drag begins.
   onDrag,
+  /// `interactive` means that the [ScrollView] will move the on-screen keyboard
+  /// relative to the pointer when a drag begins
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/scroll_view.dart:43]: Period at the end here.

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 2]
```diff
@@ -487,9 +491,28 @@ abstract class ScrollView extends StatelessWidget {
...
       );
+    } else if (keyboardDismissBehavior == ScrollViewKeyboardDismissBehavior.interactive) {
+      return Listener(
+        child: scrollableResult,
+        onPointerMove: (PointerMoveEvent details) {
+          double ypos = details.position.dy;
+
+          SystemChannels.textInput.invokeMethod<void>(
+          'TextInput.onPointerMoveForInteractiveKeyboard',
+          <String, dynamic>{'pointerY': ypos});
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/scroll_view.dart:506]: Formatting nit:

```suggestion
          SystemChannels.textInput.invokeMethod<void>(
            'TextInput.onPointerMoveForInteractiveKeyboard',
            <String, dynamic>{'pointerY': ypos},
          );
```

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 3]
```diff
@@ -487,9 +491,28 @@ abstract class ScrollView extends StatelessWidget {
...
+        },
+        onPointerUp: (PointerUpEvent details) {
+          double ypos = details.position.dy;
+
+          SystemChannels.textInput.invokeMethod<void>(
+          'TextInput.onPointerUpForInteractiveKeyboard',
+          <String, dynamic>{'pointerY': ypos});
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/scroll_view.dart:513]: Same formatting nit here.

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 4]
```diff
@@ -487,9 +491,28 @@ abstract class ScrollView extends StatelessWidget {
...
+        }
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/scroll_view.dart:514]: Trailing comma here.

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 5]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+      }
+      return null;
+    });
+
+    final TestGesture gesture = await tester.startGesture(const Offset(100.0, 100.0));
+    await gesture.moveBy(const Offset(0.0, -100.0));
+
+    expect(isPointerMoveCalled, isTrue);
+    expect(point, 0.0);
+
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:964]: Nit: Remove this extra whitespace here.

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 6]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+  });
+
+    testWidgets('Interactive Test ListView sends y coordinate on mouse up', (WidgetTester tester) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:967]: This is indented too far.

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 7]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+    ));
+
+    final Finder finder = find.byType(TextField).first;
+    final TextField textField = tester.widget(finder);
+    await tester.showKeyboard(finder);
+    expect(textField.focusNode!.hasFocus, isTrue);
+
+    bool isPointerMoveCalled = false;
+    double point = -1;
+    tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.textInput, (MethodCall methodCall) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:951]: I think you need to clean this up by calling it again with `null` after the test is finished to unset the mocked handler.

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 8]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+      if(methodCall.method == 'TextInput.onPointerMoveForInteractiveKeyboard'){
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:952]:
```suggestion
      if (methodCall.method == 'TextInput.onPointerMoveForInteractiveKeyboard') {
```

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 9]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+      }
+      return null;
+    });
+
+    final TestGesture gesture = await tester.startGesture(const Offset(100.0, 100.0));
+    await gesture.up();
+
+    expect(isPointerMoveCalled, isTrue);
+    expect(point, 100.0);
+
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:1010]: Nit: Remove the extra whitespace here.

File: packages/flutter/test/widgets/scroll_view_test.dart [hunk 10]
```diff
@@ -918,6 +918,98 @@ void main() {
...
+    ));
+
+    final Finder finder = find.byType(TextField).first;
+    final TextField textField = tester.widget(finder);
+    await tester.showKeyboard(finder);
+    expect(textField.focusNode!.hasFocus, isTrue);
+
+    bool isPointerMoveCalled = false;
+    double point = -1;
+    tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.textInput, (MethodCall methodCall) async {
```
Comment (justinmc) [inline packages/flutter/test/widgets/scroll_view_test.dart:997]: Same comment about clearing this mock.

**Review by justinmc — APPROVED** [review]
LGTM with nits 👍 .

The analyzer is failing but it might just be due to some of the formatting stuff that I commented on.

**Comment by hellohuanlin** [comment]
Let's wait for @LouiseHsu's [this PR](https://github.com/flutter/engine/pull/44554) to pass before we land this, in case there are potential test issue on the engine side.

**Review by cbracken — APPROVED** [review]
Nice! LGTM but the lints are upset.

Animated GIF of a delivery truck approaching a highway overpass, then failing to meet the height clearance requirement and crashing into it

**Comment by hellohuanlin** [comment]
Just remembered that we forget to limit this feature to exclude iPad. Let's hold this PR for now and I will take it over. @Matt2D

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 1 above)
Comment (Matt2D) [reply packages/flutter/lib/src/widgets/scroll_view.dart:43]: Always slips by me! Will add here.

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 2 above)
Comment (Matt2D) [reply packages/flutter/lib/src/widgets/scroll_view.dart:506]: Updated!

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 3 above)
Comment (Matt2D) [reply packages/flutter/lib/src/widgets/scroll_view.dart]: Missed it on this one, thank you!

File: packages/flutter/lib/src/widgets/scroll_view.dart (hunk 4 above)
Comment (Matt2D) [reply packages/flutter/lib/src/widgets/scroll_view.dart:514]: Don't know why that was missing, will add now!

File: packages/flutter/test/widgets/scroll_view_test.dart (hunk 6 above)
Comment (Matt2D) [reply packages/flutter/test/widgets/scroll_view_test.dart]: Done!

File: packages/flutter/test/widgets/scroll_view_test.dart (hunk 8 above)
Comment (Matt2D) [reply packages/flutter/test/widgets/scroll_view_test.dart:952]: Fixed in both locations now!

File: packages/flutter/test/widgets/scroll_view_test.dart (hunk 9 above)
Comment (Matt2D) [reply packages/flutter/test/widgets/scroll_view_test.dart:1009]: Gotcha!

**Comment by Piinks** [comment]
> Just remembered that we forget to limit this feature to exclude iPad. Let's hold this PR for now and I will take it over. @Matt2D

Per this comment, @hellohuanlin should we dismiss the approvals so this is not merged accidentally?

Also, for the follow up, the same change should be made in TwoDimensionalScrollView.

**Comment by hellohuanlin** [comment]
@Piinks Closing it sounds good.

**Pull request closed without merging**